ssl/statem/statem_srvr.c

```diff
@@ -3806,20 +3823,26 @@ int tls_construct_new_session_ticket(SSL *s, WPACKET *pkt)
             goto err;
         }
         s->session->ext.tick_age_add = age_add_u.age_add;
-       /*
-        * ticket_nonce is set to a single 0 byte because we only ever send a
-        * single ticket per connection. IMPORTANT: If we ever support multiple
-        * tickets per connection then this will need to be changed.
-        */
-        OPENSSL_free(s->session->ext.tick_nonce);
-        s->session->ext.tick_nonce = OPENSSL_zalloc(sizeof(char));
-        if (s->session->ext.tick_nonce == NULL) {
-            SSLfatal(s, SSL_AD_INTERNAL_ERROR,
-                     SSL_F_TLS_CONSTRUCT_NEW_SESSION_TICKET,
-                     ERR_R_MALLOC_FAILURE);
+
+        nonce = s->next_ticket_nonce;
+        memset(tick_nonce, 0, TICKET_NONCE_SIZE);
+        for (i = TICKET_NONCE_SIZE; nonce > 0 && i > 0; i--) {
+            tick_nonce[i - 1] = nonce & 0xff;
+            nonce >>= 8;
+        }
```

Comment:
There is no problem to `nonce >>= 8` is nonce is zero. Fixed-length loops are always better than variable length. Less branches is always better and fixed-length loop can be unrolled with no branches. memset would be redundant of course.

Reply:
I'm also a little bit uneasy about `nonce & 0xff`. Compiler should get it, that masked value fits the target, but some compilers don't. And in case it doesn't it would fail. appveyor seems to be happy, but older compilers burned earlier, e.g. in 3a63c0edab842af3e84ef1cad2b4eb701eece3e1.